pkg/sql/schemachanger/schemachanger_test: TestCompareLegacyAndDeclarative failed [out of bounds error in createStatsDefaultColumns]

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13767528?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13767528?buildTab=artifacts#/) on master @ [04dd7ea2c52c5ff378f6a22f58f59de1835b3feb](https://github.com/cockroachdb/cockroach/commits/04dd7ea2c52c5ff378f6a22f58f59de1835b3feb):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 73387704 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7e6a0?, 0xc012399b60?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66d83b0, 0xc015f6a2c0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:414 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x6, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:440 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x66337f8, 0xc010ec0f90}, {0x67043f0?, 0xc0102d4000}, 0x1, 0x1, 0xc8, 0xc010ec0f90?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:475 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc009464150, {0x66337f8, 0xc010ec0f90})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:259 +0x549
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc009464150, {0x66337f8, 0xc010ec0f90})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x66337f8?, 0xc010ec0f90?}, 0xc005387800?, 0xc0115ee670?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc012212b30?, 0x425e4c?}, {0x6635af8, 0xc009464150})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc012212c40, {0x6635af8?, 0xc009464150?}, {0x522e436?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc012212c40, {0x6635af8, 0xc009464150})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x66337f8?, 0xc010ec0f90?}, 0xc005387800?, 0xc0115ee670?}, {0x6635af8, 0xc009464150})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc01a8d8a80, {0x66337f8, 0xc010ec09f0})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc012cc9000?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc012cc9000, {0x66337c0?, 0xc012bc3360?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc012cc9000, {0x66337c0?, 0xc012bc3360?}, {0x6604a78?, 0xc014f9d400})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc01c222780, {0x66337c0?, 0xc012bc3360}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x669cba0?, {0x66337c0?, 0xc012bc3360?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc012507a40, {0x66337f8, 0xc0155a74a0}, 0xc004a60d10, 0xc00ac38f00, 0xc010c8c580, 0xc014f9d400, 0xc0115eeb38, 0xc012213af0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:910 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc012507a40, {0x66337f8, 0xc0155a74a0}, 0xc0115eeb38, 0xc004a60d10, 0x0?, {{0x6635af8, 0xc009464150}, 0x0}, 0xc014f9d400, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1971 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc00b6b1b48?, 0x60a999?, {0x66337f8?, 0xc0155a74a0?}, 0x0?, 0x0?, 0xc00b6b1ba0?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1686 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc004a60d10?, {0x66337f8?, 0xc0155a74a0}, 0xc0115eeb38, 0xc004a60d10, 0xc0115ee670, 0xc014f9d400, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1689 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0115ee000, {0x66337f8?, 0xc0155a74a0}, 0xc0115ee670, 0xc0155a74a0?, {0x669de80?, 0xc0191e7d60}, 0xc0191e7d60?, 0xc01fc368d8)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0115ee000, {0x66337c0, 0xc0191e7db0}, 0xc0115ee670, {0x669de80?, 0xc0191e7d60})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0115ee000, {0x66337c0, 0xc0191e7db0}, {{0x6656320, 0xc009a1d260}, {0x0, 0x0, 0x0}, {0xc00a18e600, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x66337f8?, 0xc0155a6990?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x66337f8?, {0x66337f8?, 0xc0155a6990?}, {0x6656320?, 0xc009a1d260?}, 0x3?, 0xc017dcc680?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0115ee000, {0x66337f8, 0xc0155a6990}, {{0x6656320, 0xc009a1d260}, {0x0, 0x0, 0x0}, {0xc00a18e600, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc00b6b3c48, 0xc00b6b3820, 0xc0115ee000, 0xc00b6b3760, 0xc01fe73840, 0xc01fe73830, 0xc01fe73850, 0xc00b6b3860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2221 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0115ee000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2226 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0115ee000, {0x66337f8, 0xc00f30ff50}, 0xa00c7401de0006d0?, 0xc011a16420?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2143 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 71437773
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
    comparator.go:131: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (j) USING HASH"
    comparator.go:131: Executing "CREATE TABLE t12 (i INT8)"
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35779

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13784364?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13784364?buildTab=artifacts#/) on master @ [709d30839e0dad2b5f949c03ff11ddc1a56e17db](https://github.com/cockroachdb/cockroach/commits/709d30839e0dad2b5f949c03ff11ddc1a56e17db):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 73298951 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e86640?, 0xc009af2858?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e86640?, 0xc009af2858?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e86640?, 0xc009af2858?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66e18d0, 0xc00eb56740})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:414 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x5, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:440 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x663cbd8, 0xc015953c20}, {0x670d910?, 0xc00c2fd900}, 0x1, 0x1, 0xc8, 0xc015953c20?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:475 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc00fbdc770, {0x663cbd8, 0xc015953c20})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:259 +0x549
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc00fbdc770, {0x663cbd8, 0xc015953c20})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x663cbd8?, 0xc015953c20?}, 0xc00b8df000?, 0xc0113a3970?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc020a828e0?, 0x425e4c?}, {0x663eed8, 0xc00fbdc770})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc020a829f0, {0x663eed8?, 0xc00fbdc770?}, {0x52368e6?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc020a829f0, {0x663eed8, 0xc00fbdc770})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x663cbd8?, 0xc015953c20?}, 0xc00b8df000?, 0xc0113a3970?}, {0x663eed8, 0xc00fbdc770})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc0113c8380, {0x663cbd8, 0xc015953bc0})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc0123a7200, {0x663cbd8, 0xc015953b00})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc01e34b5c0, {0x663cbd8?, 0xc015953b00?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01e34b600?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc00d513d00, {0x663cbd8?, 0xc015953b00?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc00c7dd340?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc00c7dd340, {0x663cbd8?, 0xc015953aa0?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01e1d9100?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc01e1d9100, {0x663cba0?, 0xc01123b090?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc01e1d9100, {0x663cba0?, 0xc01123b090?}, {0x660de38?, 0xc022599800})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc0152f9680, {0x663cba0?, 0xc01123b090}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x66a6080?, {0x663cba0?, 0xc01123b090?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0015fe780, {0x663cbd8, 0xc0159537d0}, 0xc00e762210, 0xc00c94be00, 0xc02e13a300, 0xc022599800, 0xc0113a3e38, 0xc020a83af0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:910 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0015fe780, {0x663cbd8, 0xc0159537d0}, 0xc0113a3e38, 0xc00e762210, 0x0?, {{0x663eed8, 0xc00fbdc770}, 0x0}, 0xc022599800, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1971 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc015ed5b48?, 0x60a9f9?, {0x663cbd8?, 0xc0159537d0?}, 0x0?, 0x0?, 0xc015ed5ba0?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1686 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc00e762210?, {0x663cbd8?, 0xc0159537d0}, 0xc0113a3e38, 0xc00e762210, 0xc0113a3970, 0xc022599800, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1689 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc0113a3300, {0x663cbd8?, 0xc0159537d0}, 0xc0113a3970, 0xc0159537d0?, {0x66a80c8?, 0xc0178b80f0}, 0xc0178b80f0?, 0xc017480758)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0113a3300, {0x663cbd8, 0xc014f42cc0}, 0xc0113a3970, {0x66a80c8?, 0xc0178b80f0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0113a3300, {0x663cbd8, 0xc014f42cc0}, {{0x665f700, 0xc011a4f2c0}, {0x0, 0x0, 0x0}, {0xc01edc09c0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x663cbd8?, 0xc00953c540?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x663cbd8?, {0x663cbd8?, 0xc00953c540?}, {0x665f700?, 0xc011a4f2c0?}, 0x3?, 0x5a?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0113a3300, {0x663cbd8, 0xc00953c540}, {{0x665f700, 0xc011a4f2c0}, {0x0, 0x0, 0x0}, {0xc01edc09c0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc015ed7c48, 0xc015ed7820, 0xc0113a3300, 0xc015ed7760, 0xc020a85840, 0xc020a85830, 0xc020a85850, 0xc015ed7860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2227 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc0113a3300)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2232 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0113a3300, {0x663cbd8, 0xc0067afce0}, 0xc0017f5620?, 0xc01ef094a0?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2149 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 70725137
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
    comparator.go:131: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (j) USING HASH"
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13805097?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13805097?buildTab=artifacts#/) on master @ [06d96981f67775e803195a8552fab9eca8f7e7a0](https://github.com/cockroachdb/cockroach/commits/06d96981f67775e803195a8552fab9eca8f7e7a0):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 67249583 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d0a0?, 0xc028ec1728?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d0a0?, 0xc028ec1728?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d0a0?, 0xc028ec1728?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66d76b0, 0xc010e28820})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:414 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x5, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:440 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x6632bd8, 0xc008dde5a0}, {0x67036f0?, 0xc011a49400}, 0x1, 0x1, 0xc8, 0xc008dde5a0?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:475 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc0092ba700, {0x6632bd8, 0xc008dde5a0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:259 +0x549
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc0092ba700, {0x6632bd8, 0xc008dde5a0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x6632bd8?, 0xc008dde5a0?}, 0xc00a90b000?, 0xc01939cc70?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc0105888e0?, 0x425e4c?}, {0x6634ed8, 0xc0092ba700})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0105889f0, {0x6634ed8?, 0xc0092ba700?}, {0x522d292?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0105889f0, {0x6634ed8, 0xc0092ba700})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x6632bd8?, 0xc008dde5a0?}, 0xc00a90b000?, 0xc01939cc70?}, {0x6634ed8, 0xc0092ba700})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc00fc2aa80, {0x6632bd8, 0xc008dde540})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc00dbca000, {0x6632bd8, 0xc008dde4e0})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc01a11e140, {0x6632bd8?, 0xc008dde4e0?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01a11e1c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc007cfe500, {0x6632bd8?, 0xc008dde4e0?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc022fcfc00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc022fcfc00, {0x6632bd8?, 0xc008dde480?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0055a6900?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc0055a6900, {0x6632ba0?, 0xc010846960?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc0055a6900, {0x6632ba0?, 0xc010846960?}, {0x6603ed8?, 0xc020e3e000})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc017a1ea00, {0x6632ba0?, 0xc010846960}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x669c040?, {0x6632ba0?, 0xc010846960?}, 0x60?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc011aa2f00, {0x6632bd8, 0xc01d647320}, 0xc0107aebb0, 0xc00c0d1360, 0xc028a86300, 0xc020e3e000, 0xc01939d138, 0xc010589af0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:910 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc011aa2f00, {0x6632bd8, 0xc01d647320}, 0xc01939d138, 0xc0107aebb0, 0x0?, {{0x6634ed8, 0xc0092ba700}, 0x0}, 0xc020e3e000, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1971 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc025649b48?, 0x60a9f9?, {0x6632bd8?, 0xc01d647320?}, 0x0?, 0x0?, 0xc025649ba0?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1686 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc0107aebb0?, {0x6632bd8?, 0xc01d647320}, 0xc01939d138, 0xc0107aebb0, 0xc01939cc70, 0xc020e3e000, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1689 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc01939c600, {0x6632bd8?, 0xc01d647320}, 0xc01939cc70, 0xc01d647320?, {0x669dfc8?, 0xc005fe93b0}, 0xc005fe93b0?, 0xc006ad8458)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc01939c600, {0x6632bd8, 0xc00acb7500}, 0xc01939cc70, {0x669dfc8?, 0xc005fe93b0})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc01939c600, {0x6632bd8, 0xc00acb7500}, {{0x6655700, 0xc0109d1860}, {0x0, 0x0, 0x0}, {0xc00631f9e0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6632bd8?, 0xc01a4fb2c0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x6632bd8?, {0x6632bd8?, 0xc01a4fb2c0?}, {0x6655700?, 0xc0109d1860?}, 0x3?, 0xc0179fe1a0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc01939c600, {0x6632bd8, 0xc01a4fb2c0}, {{0x6655700, 0xc0109d1860}, {0x0, 0x0, 0x0}, {0xc00631f9e0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc02564bc48, 0xc02564b820, 0xc01939c600, 0xc02564b760, 0xc01058b840, 0xc01058b830, 0xc01058b850, 0xc02564b860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2224 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc01939c600)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2229 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc01939c600, {0x6632bd8, 0xc01a4fa8a0}, 0xc008f85fd0?, 0xc00631fa40?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2146 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 64816918
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
    comparator.go:131: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (j) USING HASH"
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13824708?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13824708?buildTab=artifacts#/) on master @ [3c41c509a87cba7a1fd3f5cfdb0f6badb78e3704](https://github.com/cockroachdb/cockroach/commits/3c41c509a87cba7a1fd3f5cfdb0f6badb78e3704):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 74401471 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d3c0?, 0xc011c60408?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d3c0?, 0xc011c60408?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7d3c0?, 0xc011c60408?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66d64d0, 0xc00c5ea700})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:411 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x5, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:437 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x66319d8, 0xc01bbcdad0}, {0x6702510?, 0xc00593a000}, 0x1, 0x1, 0xc8, 0xc01bbcdad0?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:472 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc0082c1dc0, {0x66319d8, 0xc01bbcdad0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:256 +0x549
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc0082c1dc0, {0x66319d8, 0xc01bbcdad0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x66319d8?, 0xc01bbcdad0?}, 0xc00a83f800?, 0xc005025970?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc014cb08e0?, 0x425e4c?}, {0x6633c98, 0xc0082c1dc0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc014cb09f0, {0x6633c98?, 0xc0082c1dc0?}, {0x522d83f?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc014cb09f0, {0x6633c98, 0xc0082c1dc0})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x66319d8?, 0xc01bbcdad0?}, 0xc00a83f800?, 0xc005025970?}, {0x6633c98, 0xc0082c1dc0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc00c3a2000, {0x66319d8, 0xc01bbcda10})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc00cca2600, {0x66319d8, 0xc02272b680})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc02a42f640, {0x66319d8?, 0xc02272b680?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc02a42f6c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc01864f200, {0x66319d8?, 0xc02272b680?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc011e53a40?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc011e53a40, {0x66319d8?, 0xc01bbcd500?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc003d1b300?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc003d1b300, {0x66319a0?, 0xc0220840a0?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc003d1b300, {0x66319a0?, 0xc0220840a0?}, {0x6602cf8?, 0xc0128f2400})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc00d948f00, {0x66319a0?, 0xc0220840a0}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x669ae40?, {0x66319a0?, 0xc0220840a0?}, 0x40?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc004552000, {0x66319d8, 0xc01bbccff0}, 0xc0166a40b0, 0xc00ad95540, 0xc010d06500, 0xc0128f2400, 0xc005025e38, 0xc014cb1af0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:910 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc004552000, {0x66319d8, 0xc01bbccff0}, 0xc005025e38, 0xc0166a40b0, 0x0?, {{0x6633c98, 0xc0082c1dc0}, 0x0}, 0xc0128f2400, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1971 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc022979b48?, 0x60a9f9?, {0x66319d8?, 0xc01bbccff0?}, 0x0?, 0x0?, 0xc022979ba0?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1686 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc0166a40b0?, {0x66319d8?, 0xc01bbccff0}, 0xc005025e38, 0xc0166a40b0, 0xc005025970, 0xc0128f2400, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1689 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc005025300, {0x66319d8?, 0xc01bbccff0}, 0xc005025970, 0xc01bbccff0?, {0x669cdc8?, 0xc016267860}, 0xc016267860?, 0xc01af12f98)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc005025300, {0x66319d8, 0xc0089dd8c0}, 0xc005025970, {0x669cdc8?, 0xc016267860})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc005025300, {0x66319d8, 0xc0089dd8c0}, {{0x6654500, 0xc01674f020}, {0x0, 0x0, 0x0}, {0xc016d4b3e0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x66319d8?, 0xc01bcadc50?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x66319d8?, {0x66319d8?, 0xc01bcadc50?}, {0x6654500?, 0xc01674f020?}, 0x3?, 0xc016baf6c0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc005025300, {0x66319d8, 0xc01bcadc50}, {{0x6654500, 0xc01674f020}, {0x0, 0x0, 0x0}, {0xc016d4b3e0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc02297bc48, 0xc02297b820, 0xc005025300, 0xc02297b760, 0xc01cf4b840, 0xc01cf4b830, 0xc01cf4b850, 0xc02297b860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2224 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc005025300)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2229 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc005025300, {0x66319d8, 0xc01bbcc450}, 0xc025b01d10?, 0xc016d094a0?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2146 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 71639000
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1001, 1002); INSERT INTO t1 VALUES (1000, 1001)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13846735?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13846735?buildTab=artifacts#/) on master @ [6b88dec162f403f6d35fee3796f55bc2bffe4c38](https://github.com/cockroachdb/cockroach/commits/6b88dec162f403f6d35fee3796f55bc2bffe4c38):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 75206336 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7dc80?, 0xc01c6bcf30?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7dc80?, 0xc01c6bcf30?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7dc80?, 0xc01c6bcf30?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66d6ff0, 0xc01ed65d20})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:411 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x5, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:437 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x66324f8, 0xc020188630}, {0x6703030?, 0xc00a9cb400}, 0x1, 0x1, 0xc8, 0xc020188630?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:472 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc01c2d2310, {0x66324f8, 0xc020188630})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:256 +0x549
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc01c2d2310, {0x66324f8, 0xc020188630})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:128 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x66324f8?, 0xc020188630?}, 0xc0053d7800?, 0xc01f39b2f0?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:115 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc0212508e0?, 0x425e4c?}, {0x66347b8, 0xc01c2d2310})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0212509f0, {0x66347b8?, 0xc01c2d2310?}, {0x522e0ff?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0212509f0, {0x66347b8, 0xc01c2d2310})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x66324f8?, 0xc020188630?}, 0xc0053d7800?, 0xc01f39b2f0?}, {0x66347b8, 0xc01c2d2310})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc01ff85880, {0x66324f8, 0xc020188570})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc00ca00c00, {0x66324f8, 0xc020188510})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc01f839ec0, {0x66324f8?, 0xc020188510?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01f839f00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc01cecd500, {0x66324f8?, 0xc020188510?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc00602f6c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc00602f6c0, {0x66324f8?, 0xc0201884b0?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01610bc00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc01610bc00, {0x66324c0?, 0xc011d09180?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc01610bc00, {0x66324c0?, 0xc011d09180?}, {0x6603818?, 0xc0194c2400})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc01cf35400, {0x66324c0?, 0xc011d09180}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x669b960?, {0x66324c0?, 0xc011d09180?}, 0x80?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0160c28c0, {0x66324f8, 0xc020188210}, 0xc01ba82580, 0xc0201ec280, 0xc011795700, 0xc0194c2400, 0xc01f39b7b8, 0xc021251af0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:910 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc0160c28c0, {0x66324f8, 0xc020188210}, 0xc01f39b7b8, 0xc01ba82580, 0x0?, {{0x66347b8, 0xc01c2d2310}, 0x0}, 0xc0194c2400, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1971 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc02c245b48?, 0x60a9f9?, {0x66324f8?, 0xc020188210?}, 0x0?, 0x0?, 0xc02c245ba0?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1686 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc01ba82580?, {0x66324f8?, 0xc020188210}, 0xc01f39b7b8, 0xc01ba82580, 0xc01f39b2f0, 0xc0194c2400, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1689 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc01f39ac80, {0x66324f8?, 0xc020188210}, 0xc01f39b2f0, 0xc020188210?, {0x669d8e8?, 0xc020b45130}, 0xc020b45130?, 0xc01b7b8458)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc01f39ac80, {0x66324f8, 0xc0150cdaa0}, 0xc01f39b2f0, {0x669d8e8?, 0xc020b45130})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc01f39ac80, {0x66324f8, 0xc0150cdaa0}, {{0x6655020, 0xc02028fa40}, {0x0, 0x0, 0x0}, {0xc0202ae420, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x66324f8?, 0xc0016bda10?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x66324f8?, {0x66324f8?, 0xc0016bda10?}, {0x6655020?, 0xc02028fa40?}, 0x3?, 0xc01f47b520?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc01f39ac80, {0x66324f8, 0xc0016bda10}, {{0x6655020, 0xc02028fa40}, {0x0, 0x0, 0x0}, {0xc0202ae420, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc02c247c48, 0xc02c247820, 0xc01f39ac80, 0xc02c247760, 0xc021649840, 0xc021649830, 0xc021649850, 0xc02c247860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2224 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc01f39ac80)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2229 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc01f39ac80, {0x66324f8, 0xc01f4edef0}, 0xc00597e450?, 0xc020302a20?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2146 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 72946070
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
    comparator.go:131: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (j) USING HASH"
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13868131?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13868131?buildTab=artifacts#/) on master @ [4d31467eef5861d7c2261b0108f9df6a1bbf13a9](https://github.com/cockroachdb/cockroach/commits/4d31467eef5861d7c2261b0108f9df6a1bbf13a9):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 73593021 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7e8c0?, 0xc01013d008?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7e8c0?, 0xc01013d008?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e7e8c0?, 0xc01013d008?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66d8470, 0xc01c1ce7c0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:398 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x4, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:424 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x6633958, 0xc01ec54120}, {0x67044b0?, 0xc006356a00}, 0x1, 0x1, 0xc8, 0xc01ec54120?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:459 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc00b6e29a0, {0x6633958, 0xc01ec54120})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:243 +0x589
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc00b6e29a0, {0x6633958, 0xc01ec54120})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:127 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x6633958?, 0xc01ec54120?}, 0xc0083ac000?, 0xc01acd6670?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:114 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc00cb228d0?, 0x425e4c?}, {0x6635c18, 0xc00b6e29a0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc00cb229e0, {0x6635c18?, 0xc00b6e29a0?}, {0x522ec8d?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc00cb229e0, {0x6635c18, 0xc00b6e29a0})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x6633958?, 0xc01ec54120?}, 0xc0083ac000?, 0xc01acd6670?}, {0x6635c18, 0xc00b6e29a0})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc01abd7500, {0x6633958, 0xc01ec540c0})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc01c7b6600, {0x6633958, 0xc01ec54060})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc0140892c0, {0x6633958?, 0xc01ec54060?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc014089300?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc02035b300, {0x6633958?, 0xc01ec54060?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc000cd0e00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc000cd0e00, {0x6633958?, 0xc01ec54000?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc005100700?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc005100700, {0x6633920?, 0xc013eaf810?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc005100700, {0x6633920?, 0xc013eaf810?}, {0x6604d78?, 0xc0104bac00})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc00f5bc500, {0x6633920?, 0xc013eaf810}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x669cde0?, {0x6633920?, 0xc013eaf810?}, 0x60?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc016bf68c0, {0x6633958, 0xc00d57b770}, 0xc00fd5d340, 0xc00cf1f860, 0xc001d5a080, 0xc0104bac00, 0xc01acd6b38, 0xc00cb23ae0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:909 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc016bf68c0, {0x6633958, 0xc00d57b770}, 0xc01acd6b38, 0xc00fd5d340, 0x0?, {{0x6635c18, 0xc00b6e29a0}, 0x0}, 0xc0104bac00, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1970 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc0175c1b38?, 0x60a9f9?, {0x6633958?, 0xc00d57b770?}, 0x0?, 0x0?, 0xc0175c1b90?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1685 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc00fd5d340?, {0x6633958?, 0xc00d57b770}, 0xc01acd6b38, 0xc00fd5d340, 0xc01acd6670, 0xc0104bac00, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1688 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc01acd6000, {0x6633958?, 0xc00d57b770}, 0xc01acd6670, 0xc00d57b770?, {0x669ed68?, 0xc006654b40}, 0xc006654b40?, 0xc007a8f358)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc01acd6000, {0x6633958, 0xc0167dd500}, 0xc01acd6670, {0x669ed68?, 0xc006654b40})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc01acd6000, {0x6633958, 0xc0167dd500}, {{0x6656480, 0xc00bca74a0}, {0x0, 0x0, 0x0}, {0xc014112ba0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6633958?, 0xc014d53b60?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x6633958?, {0x6633958?, 0xc014d53b60?}, {0x6656480?, 0xc00bca74a0?}, 0x3?, 0x5a?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc01acd6000, {0x6633958, 0xc014d53b60}, {{0x6656480, 0xc00bca74a0}, {0x0, 0x0, 0x0}, {0xc014112ba0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc0175c3c48, 0xc0175c3820, 0xc01acd6000, 0xc0175c3760, 0xc00cb25840, 0xc00cb25830, 0xc00cb25850, 0xc0175c3860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2224 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc01acd6000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2229 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc01acd6000, {0x6633958, 0xc014d53200}, 0xc0054ebfd0?, 0xc006799080?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2146 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 71295732
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1001, 1002); INSERT INTO t1 VALUES (1000, 1001)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13888107?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13888107?buildTab=artifacts#/) on master @ [c5ae229b316131bcda31892738ae0e40172e11e7](https://github.com/cockroachdb/cockroach/commits/c5ae229b316131bcda31892738ae0e40172e11e7):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 72680267 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e81c00?, 0xc00ea9fe78?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e81c00?, 0xc00ea9fe78?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e81c00?, 0xc00ea9fe78?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66dc710, 0xc029074380})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:398 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x4, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:424 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x6637c58, 0xc01f25c900}, {0x6708750?, 0xc007c41400}, 0x1, 0x1, 0xc8, 0xc01f25c900?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:459 +0x58a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc00aadbe30, {0x6637c58, 0xc01f25c900})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:243 +0x589
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc00aadbe30, {0x6637c58, 0xc01f25c900})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:127 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x6637c58?, 0xc01f25c900?}, 0xc005786000?, 0xc009799978?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:114 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc016b9a8d0?, 0x425e4c?}, {0x6639f18, 0xc00aadbe30})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc016b9a9e0, {0x6639f18?, 0xc00aadbe30?}, {0x52328ab?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc016b9a9e0, {0x6639f18, 0xc00aadbe30})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x6637c58?, 0xc01f25c900?}, 0xc005786000?, 0xc009799978?}, {0x6639f18, 0xc00aadbe30})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc019842a80, {0x6637c58, 0xc01f25c840})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc0175e5200, {0x6637c58, 0xc01f25c7e0})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc01b41f880, {0x6637c58?, 0xc01f25c7e0?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01b41f8c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc00140d800, {0x6637c58?, 0xc01f25c7e0?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0125528c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc0125528c0, {0x6637c58?, 0xc01f25c780?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc028e9bd00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc028e9bd00, {0x6637c20?, 0xc010aae0f0?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc028e9bd00, {0x6637c20?, 0xc010aae0f0?}, {0x6609018?, 0xc023755c00})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc01f1eac80, {0x6637c20?, 0xc010aae0f0}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x66a0fc0?, {0x6637c20?, 0xc010aae0f0?}, 0x40?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc014782000, {0x6637c58, 0xc016df7ce0}, 0xc0114dd6b0, 0xc02951cb40, 0xc01fb49680, 0xc023755c00, 0xc009799e48, 0xc016b9bae0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:909 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc014782000, {0x6637c58, 0xc016df7ce0}, 0xc009799e48, 0xc0114dd6b0, 0x0?, {{0x6639f18, 0xc00aadbe30}, 0x0}, 0xc023755c00, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1970 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc01c5cbb38?, 0x4ad92e0?, {0x6637c58?, 0xc016df7ce0?}, 0x0?, 0xc01c5cbb68?, 0x492949?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1685 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc0114dd6b0?, {0x6637c58?, 0xc016df7ce0}, 0xc009799e48, 0xc0114dd6b0, 0xc009799978, 0xc023755c00, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1688 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc009799300, {0x6637c58?, 0xc016df7ce0}, 0xc009799978, 0xc016df7ce0?, {0x66a3008?, 0xc015b37b80}, 0xc015b37b80?, 0xc00d3a8e18)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2420 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc009799300, {0x6637c58, 0xc016df7770}, 0xc009799978, {0x66a3008?, 0xc015b37b80})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1949 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc009799300, {0x6637c58, 0xc016df7770}, {{0x665a780, 0xc017199380}, {0x0, 0x0, 0x0}, {0xc013429ce0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x42c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6637c58?, 0xc016df70b0?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x6637c58?, {0x6637c58?, 0xc016df70b0?}, {0x665a780?, 0xc017199380?}, 0x3?, 0x5a?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3427 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc009799300, {0x6637c58, 0xc016df70b0}, {{0x665a780, 0xc017199380}, {0x0, 0x0, 0x0}, {0xc013429ce0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc01c5cdc48, 0xc01c5cd820, 0xc009799300, 0xc01c5cd760, 0xc016b9d840, 0xc016b9d830, 0xc016b9d850, 0xc01c5cd860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2236 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc009799300)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2241 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc009799300, {0x6637c58, 0xc016df6750}, 0xc01be0af50?, 0xc0088caa20?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2158 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 69944787
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1001, 1002); INSERT INTO t1 VALUES (1000, 1001)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/sql/schemachanger/schemachanger_test.TestCompareLegacyAndDeclarative [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13906792?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13906792?buildTab=artifacts#/) on master @ [bf5ffc4da2b33e3b9ed7bb3667e9ab0710256e4e](https://github.com/cockroachdb/cockroach/commits/bf5ffc4da2b33e3b9ed7bb3667e9ab0710256e4e):

Fatal error:

```
panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3 [recovered]
	panic: runtime error: index out of range [3] with length 3

```
Stack: 

```
goroutine 75040628 [running]:
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e89520?, 0xc014a07c68?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e89520?, 0xc014a07c68?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x3bb
panic({0x4e89520?, 0xc014a07c68?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func1({0x66e5eb0, 0xc017c98820})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:398 +0x8d
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns.func3(0x5, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:424 +0x64
github.com/cockroachdb/cockroach/pkg/sql.createStatsDefaultColumns({0x6641378, 0xc0217681e0}, {0x6711f08?, 0xc00952b400}, 0x1, 0x1, 0xc8, 0xc0217681e0?)
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:459 +0x66a
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).makeJobRecord(0xc00a450d90, {0x6641378, 0xc0217681e0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:243 +0x589
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).runJob(0xc00a450d90, {0x6641378, 0xc0217681e0})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:127 +0x37
github.com/cockroachdb/cockroach/pkg/sql.(*createStatsNode).startExec(0x0?, {{0x6641378?, 0xc0217681e0?}, 0xc012d4e000?, 0xc004e152f8?})
	github.com/cockroachdb/cockroach/pkg/sql/create_stats.go:114 +0x74
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc00980c8b0?, 0x425e4c?}, {0x6643638, 0xc00a450d90})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:520 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x38
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc00980c9c0, {0x6643638?, 0xc00a450d90?}, {0x523a1fb?, 0x11?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x10c2
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc00980c9c0, {0x6643638, 0xc00a450d90})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf0
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x6641378?, 0xc0217681e0?}, 0xc012d4e000?, 0xc004e152f8?}, {0x6643638, 0xc00a450d90})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x10b
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc01a0db180, {0x6641378, 0xc021768150})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:175 +0x125
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc010089200, {0x6641378, 0xc020d88930})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xe8
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc01cedf580, {0x6641378?, 0xc020d88930?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x49
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(...)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc01cedf5c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc0237b4600, {0x6641378?, 0xc020d88930?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x77
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc0172f21c0?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0xc0172f21c0, {0x6641378?, 0xc020d888d0?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x89
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc00b09bc00?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x57
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0xc00b09bc00, {0x6641340?, 0xc01c1e63c0?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x7b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc00b09bc00, {0x6641340?, 0xc01c1e63c0?}, {0x6612738?, 0xc006eb3400})
	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:722 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc0144e2280, {0x6641340?, 0xc01c1e63c0}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x222
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x66aa760?, {0x6641340?, 0xc01c1e63c0?}, 0x60?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x225
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc00f7fd040, {0x6641378, 0xc020d88360}, 0xc00d0480b0, 0xc016195d60, 0xc001641300, 0xc006eb3400, 0xc004e157c8, 0xc00980dac0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:909 +0xc87
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc00f7fd040, {0x6641378, 0xc020d88360}, 0xc004e157c8, 0xc00d0480b0, 0x9a02748?, {{0x6643638, 0xc00a450d90}, 0x0}, 0xc006eb3400, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1970 +0x215
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0x4240be?, 0x7f638eea7c00?, {0x6641378?, 0xc020d88360?}, 0x2e?, 0xc019ab1b28?, 0x431c76?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1685 +0xbb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc00d0480b0?, {0x6641378?, 0xc020d88360}, 0xc004e157c8, 0xc00d0480b0, 0xc004e152f8, 0xc006eb3400, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1688 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004e14c80, {0x6641378?, 0xc020d88360}, 0xc004e152f8, 0xc020d88360?, {0x66ac7a8?, 0xc01b7fe320}, 0xc01b7fe320?, 0xc01d9f8a58)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2439 +0x45c
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004e14c80, {0x6641378, 0xc0015b4f30}, 0xc004e152f8, {0x66ac7a8?, 0xc01b7fe320})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1968 +0x1312
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004e14c80, {0x6641378, 0xc0015b4f30}, {{0x6663ea0, 0xc00eeb1020}, {0x0, 0x0, 0x0}, {0xc009bbd1a0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1144 +0x465d
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6641378?, 0xc0125ede00?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:146 +0xb0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x6641378?, {0x6641378?, 0xc0125ede00?}, {0x6663ea0?, 0xc00eeb1020?}, 0x3?, 0xc0061a4b60?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3446 +0x313
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004e14c80, {0x6641378, 0xc0125ede00}, {{0x6663ea0, 0xc00eeb1020}, {0x0, 0x0, 0x0}, {0xc009bbd1a0, 0x5a}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x4a5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc019ab3c48, 0xc019ab3820, 0xc004e14c80, 0xc019ab3760, 0xc00980f840, 0xc00980f830, 0xc00980f850, 0xc019ab3860)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2244 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004e14c80)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2249 +0x336
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004e14c80, {0x6641378, 0xc015e18990}, 0xc00a772fd0?, 0xc00eec53e0?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2166 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0xa5
created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 72162147
	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x176
```
<details><summary>Log preceding fatal error</summary>
<p>

```
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1001, 1002); INSERT INTO t1 VALUES (1000, 1001)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t3 (i INT8 NOT NULL); INSERT INTO t3 SELECT generate_series(1, 1000)"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ALTER PRIMARY KEY USING COLUMNS (i); INSERT INTO t3 VALUES (1001); COMMIT TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 NOT NULL); BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 ADD PRIMARY KEY (i)"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "SELECT 1 / 0"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t2"
    comparator.go:131: Executing "INSERT INTO t2 VALUES (1002, 1003); INSERT INTO t1 VALUES (1001, 1002)"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "DROP TABLE IF EXISTS t3; CREATE TABLE t3 (i INT8 PRIMARY KEY)"
    comparator.go:131: Executing "BEGIN TRANSACTION"
    comparator.go:131: Executing "ALTER TABLE t3 DROP CONSTRAINT t3_pkey"
    comparator.go:131: Executing "DELETE FROM t3 WHERE i = 1"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; ALTER TABLE t3 ADD COLUMN j INT8 CREATE FAMILY"
    comparator.go:131: Executing "ROLLBACK TRANSACTION"
    comparator.go:131: Executing "BEGIN TRANSACTION; SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "RELEASE SAVEPOINT cockroach_restart"
    comparator.go:131: Executing "SELECT 1"
    comparator.go:131: Executing "COMMIT TRANSACTION"
    comparator.go:131: Executing "CREATE TABLE t11 (i INT8 NOT NULL); ALTER TABLE t11 ALTER PRIMARY KEY USING COLUMNS (i)"
    comparator.go:131: Executing "CREATE SEQUENCE s"
    comparator.go:131: Executing "CREATE TABLE t4 (i INT8 CHECK (i > nextval('s')) CHECK (i > 0), CONSTRAINT ck_i CHECK (i > nextval('s'::REGCLASS)), CONSTRAINT ck_i2 CHECK (i > 0))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD CHECK (i > nextval('s'))"
    comparator.go:131: Executing "ALTER TABLE t4 ADD COLUMN j INT8 CHECK (j > 0) CHECK ((i + j) > nextval('s'))"
    comparator.go:131: Executing "CREATE TABLE t5 (i INT8 NOT NULL)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" (for legacy) and to "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" with modifications: "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i); COMMIT TRANSACTION; ALTER TABLE t5 DROP COLUMN IF EXISTS rowid" for legacy; "ALTER TABLE t5 ALTER PRIMARY KEY USING COLUMNS (i)" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET use_declarative_schema_changer = off" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET use_declarative_schema_changer = off" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "SET CLUSTER SETTING \"sql.schema.force_declarative_statements\" = \"-CREATE SCHEMA, +CREATE SEQUENCE\"" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t6 (i INT8 PRIMARY KEY, j INT8 NOT NULL, k INT8 NOT NULL)"
    comparator.go:131: Executing "ALTER TABLE t6 DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i"
    comparator.go:131: Executing "ALTER TABLE t6 ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k"
    comparator.go:131: Executing "SET experimental_enable_temp_tables = true"
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:245: Comparator testing framework modifies line "CREATE TEMPORARY TABLE t7 (i INT8)" to "" (for legacy) and to "" (for declarative)
    comparator.go:133: Executing "CREATE TEMPORARY TABLE t7 (i INT8)" with modifications: "" for legacy; "" for declarative
    comparator.go:131: Executing "CREATE TABLE t10 (i INT8 NOT NULL, j INT8 NOT NULL, k INT8 NOT NULL, PRIMARY KEY (i, k) USING HASH WITH (bucket_count = 3))"
    comparator.go:131: Executing "INSERT INTO t10 VALUES (0, 1, 2)"
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH; COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_3" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k) USING HASH" for declarative
    comparator.go:245: Comparator testing framework modifies line "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" (for legacy) and to "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" (for declarative)
    comparator.go:133: Executing "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" with modifications: "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k); COMMIT TRANSACTION; ALTER TABLE t10 DROP COLUMN IF EXISTS crdb_internal_i_k_shard_16" for legacy; "ALTER TABLE t10 ALTER PRIMARY KEY USING COLUMNS (i, k)" for declarative
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss,deadlock</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCompareLegacyAndDeclarative.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/create_stats.go
+++ b/pkg/sql/create_stats.go
@@ -420,6 +420,12 @@ func createStatsDefaultColumns(
 			return err
 		}
 
+		// There shouldn't be any non-public columns, but defensively skip over them
+		// if there are.
+		if !col.Public() {
+			return nil
+		}
+
 		// Skip unsupported virtual computed columns.
 		if isUnsupportedVirtual(col) {
 			return nil
@@ -466,9 +472,30 @@ func createStatsDefaultColumns(
 			continue
 		}
 
-		colIDs := make([]descpb.ColumnID, i+1)
+		colIDs := make([]descpb.ColumnID, 0, i+1)
 		for j := 0; j <= i; j++ {
-			colIDs[j] = desc.GetPrimaryIndex().GetKeyColumnID(j)
+			col, err := catalog.MustFindColumnByID(desc, desc.GetPrimaryIndex().GetKeyColumnID(j))
+			if err != nil {
+				return nil, err
+			}
+
+			// There shouldn't be any non-public columns, but defensively skip over
+			// them if there are.
+			if !col.Public() {
+				continue
+			}
+
+			// Skip unsupported virtual computed columns.
+			if isUnsupportedVirtual(col) {
+				continue
+			}
+			colIDs = append(colIDs, col.GetID())
+		}
+
+		// Do not attempt to create multi-column stats with < 2 columns. This can
+		// happen when an index contains only virtual computed columns.
+		if len(colIDs) < 2 {
+			continue
 		}
 
 		// Remember the requested stats so we don't request duplicates.
@@ -503,6 +530,13 @@ func createStatsDefaultColumns(
 				if err != nil {
 					return nil, err
 				}
+
+				// There shouldn't be any non-public columns, but defensively skip them
+				// if there are.
+				if !col.Public() {
+					continue
+				}
+
 				// Skip unsupported virtual computed columns.
 				if isUnsupportedVirtual(col) {
 					continue
@@ -510,9 +544,9 @@ func createStatsDefaultColumns(
 				colIDs = append(colIDs, col.GetID())
 			}
 
-			// Do not attempt to create multi-column stats with no columns. This
-			// can happen when an index contains only virtual computed columns.
-			if len(colIDs) == 0 {
+			// Do not attempt to create multi-column stats with < 2 columns. This can
+			// happen when an index contains only virtual computed columns.
+			if len(colIDs) < 2 {
 				continue
 			}
 

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -2670,3 +2670,38 @@ upper_bound            range_rows  distinct_range_rows  equal_rows
 
 statement ok
 RESET enable_create_stats_using_extremes
+
+# Regression test for #118537. Do not create stats on non-public mutation
+# columns.
+statement ok
+CREATE TABLE t118537 (
+  a INT,
+  PRIMARY KEY (a) USING HASH WITH (bucket_count = 3)
+)
+
+statement ok
+INSERT INTO t118537 SELECT generate_series(0, 9)
+
+statement ok
+SET CLUSTER SETTING jobs.debug.pausepoints = 'newschemachanger.before.exec'
+
+skipif config local-legacy-schema-changer
+statement error job \d+ was paused before it completed with reason: pause point "newschemachanger.before.exec" hit
+ALTER TABLE t118537 ALTER PRIMARY KEY USING COLUMNS (a) USING HASH
+
+statement ok
+CREATE STATISTICS mutation FROM t118537
+
+query TTIB colnames
+SELECT statistics_name, column_names, row_count, histogram_id IS NOT NULL AS has_histogram
+FROM [SHOW STATISTICS FOR TABLE t118537]
+ORDER BY statistics_name, column_names::STRING
+----
+statistics_name  column_names  row_count  has_histogram
+mutation         {a}           10         true
+
+statement ok
+SET CLUSTER SETTING jobs.debug.pausepoints = ''
+
+statement ok
+RESUME JOB (SELECT job_id FROM crdb_internal.jobs WHERE description LIKE 'ALTER TABLE %t118537 ALTER PRIMARY KEY USING COLUMNS (a) USING HASH' AND status = 'paused' FETCH FIRST 1 ROWS ONLY)

